Source/WTF/wtf/text/StringImpl.cpp

```diff
@@ -1416,9 +1416,11 @@ template<typename CharacterType> inline bool equalInternal(const StringImpl* a,
 
     if (a->length() != length)
         return false;
+    if (!length)
+        return true;
     if (a->is8Bit())
-        return equal(a->characters8(), b, length);
-    return equal(a->characters16(), b, length);
+        return *a->characters8() == *b && equal(a->characters8() + 1, b + 1, length - 1);
+    return *a->characters16() == *b && equal(a->characters16() + 1, b + 1, length - 1);
```

Comment:
How did you decide to put the optimization here rather than in the inline `equal(a, b, length)` functions?

Reply:
This is the place where we were doing the is8Bit() check already and I need access to the first character. Given this, it seems like a good place to add the check.

Reply:
Seems to me like this would be even better inside the highly-optimized `ALWAYS_INLINE` `equal` functions, not here just before calling them.

Reply:
The function that is really hot is:
```
WTF_EXPORT_PRIVATE bool equal(const StringImpl*, const LChar*, unsigned); // In StringImpl.h
```

It is not inline. It is defined in the cpp like so:
```
bool equal(const StringImpl* a, const LChar* b, unsigned length)
{
    return equalInternal(a, b, length);
}
```

This is why I modified `equalInternal()`.

```
Sample Count, Samples %, Normalized CPU %, Symbol
483, 0.2%, 0.1%, WTF::equal(WTF::StringImpl const*, unsigned char const*, unsigned int) (in JavaScriptCore)
217, 0.1%, 0.0%,     WTF::AtomStringImpl::add(unsigned char const*, unsigned int) (in JavaScriptCore)
146, 0.1%, 0.0%,     JSC::jsonProtoFuncParse(JSC::JSGlobalObject*, JSC::CallFrame*) (in JavaScriptCore)
33, 0.0%, 0.0%,     WebCore::HTMLFastPathParser<unsigned char>::parseAttributes(WebCore::Element&) (in WebCore)
```

Reply:
Is your suggestion that we try to make part of this function inline and do the quick checks in the inline portion?

Reply:
No.

`StringCommon.h` has `ALWAYS_INLINE` functions `equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)`. I am suggesting that we do the quick checks in those functions instead of in `equalInternal`.

I don’t think you need to do that in this patch. I am saying that I believe putting the optimizations there _might_ give us even better results, because other call sites likely would get similar benefit. But it’s also possible it might not work as well.

Separately, maybe it is worth exploring a more-inlined version, what you thought I might mean. The fact that we see this function call at all on the profile, given how simple the algorithm is, I think often does mean there is a possible improvement.

Reply:
I think there is a misunderstanding. If you look at the trace (pasted above) and the corresponding code, you'll see that the AtomStringTable code calls `equal(const StringImpl* a, const LChar* b, unsigned length)` directly for example.

Looks to me that moving the checks to the function you suggest may help for other workloads but won't impact the hot ones on Speedometer?

Reply:
`equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)` are other variants, not using the code path I am modifying in this patch. They are not the ones I see as super high on Speedometer.

I can investigate those separately but unless I am missing something they are not using the code I am modifying and not related to the Speedometer samples I am working on reducing.

Reply:
> Looks to me that moving the checks to the function you suggest may help for other workloads but won't impact the hot ones on Speedometer?

Yes. Sorry if I accidentally seemed to imply otherwise.

No further impact on Speedometer might still mean improvement in other cases. Or moving it might even ruin the Speedometer improvement. But it seems likely to me that most string comparisons could benefit from a fast out for first character mismatches.

I’m **glad** you are fully focused on the actual measurable Speedometer impact, but I was thinking slightly more broadly when making my suggestion.

Reply:
> `equal(const LChar*, const LChar*, unsigned)` and `equal(const UChar*, const UChar*, unsigned)` are other variants, not using the code path I am modifying in this patch. They are not the ones I see as super high on Speedometer.

Wait, what? I am pretty sure that they **are** the functions called in the last two lines of `equalInternal`.

Reply:
OHHHH, indeed :) I got confused, I thought you were talking about one of the callers but those are actually callees.
I see your point better now. I'll do some more A/B testing.

Reply:
And now I realize that if your approach works out, we'd need a single patch instead of at least two:
https://github.com/WebKit/WebKit/pull/12239

Reply:
Doing the check in the ALWAYS_INLINE equal() didn't measure as a progression so I'll land this version for now.